File: backend/app/schemas/credit_application.py
"""
Pydantic schemas for in-CRM lead credit applications.
All fields optional — nothing blocks save or submit.
"""
from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ResidenceSection(BaseModel):
    street: Optional[str] = None
    apt: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    zip: Optional[str] = None
    years: Optional[int] = None
    months: Optional[int] = None
    home_phone: Optional[str] = None
    cell_phone: Optional[str] = None
    monthly_rent_mortgage: Optional[float] = None
    residential_status: Optional[str] = None
    landlord_name: Optional[str] = None
    landlord_phone: Optional[str] = None


class EmploymentSection(BaseModel):
    employer_name: Optional[str] = None
    gross_monthly_salary: Optional[float] = None
    address: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    zip: Optional[str] = None
    work_phone: Optional[str] = None
    years: Optional[int] = None
    months: Optional[int] = None
    occupation: Optional[str] = None


class ApplicantSection(BaseModel):
    full_name: Optional[str] = None
    date_of_birth: Optional[str] = None
    ssn: Optional[str] = Field(None, description="Write-only; stored encrypted")
    ssn_masked: Optional[str] = Field(None, description="Read-only masked SSN")
    current_residence: Optional[ResidenceSection] = None
    previous_residence: Optional[ResidenceSection] = None
    current_employment: Optional[EmploymentSection] = None
    previous_employment: Optional[EmploymentSection] = None
    secondary_employment: Optional[EmploymentSection] = None


class OtherIncomeSection(BaseModel):
    gross_monthly_other_income: Optional[float] = None
    source: Optional[str] = None
    belongs_to: Optional[str] = None


class ReferenceSection(BaseModel):
    name: Optional[str] = None
    phone: Optional[str] = None
    address: Optional[str] = None
    relationship: Optional[str] = None


class BankReferenceSection(BaseModel):
    bank_name: Optional[str] = None
    account_type: Optional[str] = None


class AuthorizationSection(BaseModel):
    applicant_signature: Optional[str] = None
    applicant_signature_date: Optional[str] = None
    applicant_dl_number: Optional[str] = None
    joint_signature: Optional[str] = None
    joint_signature_date: Optional[str] = None
    joint_dl_number: Optional[str] = None


class VehicleSection(BaseModel):
    condition: Optional[str] = None
    year: Optional[str] = None
    make: Optional[str] = None
    model: Optional[str] = None
    body_style: Optional[str] = None
    mileage: Optional[str] = None
    vin: Optional[str] = None


class TradeInSection(BaseModel):
    year: Optional[str] = None
    make: Optional[str] = None
    model: Optional[str] = None
    body_style: Optional[str] = None
    lienholder: Optional[str] = None
    allowance: Optional[float] = None
    payoff: Optional[float] = None


class DealerSection(BaseModel):
    vehicle: Optional[VehicleSection] = None
    trade_in: Optional[TradeInSection] = None
    cash_selling_price: Optional[float] = None
    net_trade: Optional[float] = None
    cash_down: Optional[float] = None
    products_and_fees: Optional[float] = None
    amount_financed: Optional[float] = None
    term_months: Optional[int] = None
    apr: Optional[float] = None


class CreditApplicationUpdate(BaseModel):
    transaction_type: Optional[Literal["credit_sale", "lease"]] = None
    application_type: Optional[Literal["individual", "joint", "business"]] = None
    app_number: Optional[str] = None
    applicant_a: Optional[ApplicantSection] = None
    applicant_b: Optional[ApplicantSection] = None
    other_income: Optional[OtherIncomeSection] = None
    reference: Optional[ReferenceSection] = None
    bank_reference: Optional[BankReferenceSection] = None
    authorization: Optional[AuthorizationSection] = None
    dealer_section: Optional[DealerSection] = None
    revert_to_draft: Optional[bool] = Field(
        False, description="If true and status is submitted, revert to draft for editing"
    )


class CreditApplicationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    lead_id: UUID
    dealership_id: Optional[UUID] = None
    status: str
    transaction_type: Optional[str] = None
    application_type: Optional[str] = None
    app_number: Optional[str] = None
    applicant_a: Optional[ApplicantSection] = None
    applicant_b: Optional[ApplicantSection] = None
    other_income: Optional[OtherIncomeSection] = None
    reference: Optional[ReferenceSection] = None
    bank_reference: Optional[BankReferenceSection] = None
    authorization: Optional[AuthorizationSection] = None
    dealer_section: Optional[DealerSection] = None
    created_by: Optional[UUID] = None
    updated_by: Optional[UUID] = None
    submitted_by: Optional[UUID] = None
    submitted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    updated_by_name: Optional[str] = None
    submitted_by_name: Optional[str] = None

File: frontend/src/types/credit-application.ts
export type CreditApplicationStatus = "draft" | "submitted"
export type TransactionType = "credit_sale" | "lease"
export type ApplicationType = "individual" | "joint" | "business"

export interface ResidenceSection {
  street?: string | null
  apt?: string | null
  city?: string | null
  state?: string | null
  zip?: string | null
  years?: number | null
  months?: number | null
  home_phone?: string | null
  cell_phone?: string | null
  monthly_rent_mortgage?: number | null
  residential_status?: string | null
  landlord_name?: string | null
  landlord_phone?: string | null
}

export interface EmploymentSection {
  employer_name?: string | null
  gross_monthly_salary?: number | null
  address?: string | null
  city?: string | null
  state?: string | null
  zip?: string | null
  work_phone?: string | null
  years?: number | null
  months?: number | null
  occupation?: string | null
}

export interface ApplicantSection {
  full_name?: string | null
  date_of_birth?: string | null
  ssn?: string | null
  ssn_masked?: string | null
  current_residence?: ResidenceSection | null
  previous_residence?: ResidenceSection | null
  current_employment?: EmploymentSection | null
  previous_employment?: EmploymentSection | null
  secondary_employment?: EmploymentSection | null
}

export interface OtherIncomeSection {
  gross_monthly_other_income?: number | null
  source?: string | null
  belongs_to?: string | null
}

export interface ReferenceSection {
  name?: string | null
  phone?: string | null
  address?: string | null
  relationship?: string | null
}

export interface BankReferenceSection {
  bank_name?: string | null
  account_type?: string | null
}

export interface AuthorizationSection {
  applicant_signature?: string | null
  applicant_signature_date?: string | null
  applicant_dl_number?: string | null
  joint_signature?: string | null
  joint_signature_date?: string | null
  joint_dl_number?: string | null
}

export interface VehicleSection {
  condition?: string | null
  year?: string | null
  make?: string | null
  model?: string | null
  body_style?: string | null
  mileage?: string | null
  vin?: string | null
}

export interface TradeInSection {
  year?: string | null
  make?: string | null
  model?: string | null
  body_style?: string | null
  lienholder?: string | null
  allowance?: number | null
  payoff?: number | null
}

export interface DealerSection {
  vehicle?: VehicleSection | null
  trade_in?: TradeInSection | null
  cash_selling_price?: number | null
  net_trade?: number | null
  cash_down?: number | null
  products_and_fees?: number | null
  amount_financed?: number | null
  term_months?: number | null
  apr?: number | null
}

export interface CreditApplication {
  id: string
  lead_id: string
  dealership_id?: string | null
  status: CreditApplicationStatus
  transaction_type?: TransactionType | null
  application_type?: ApplicationType | null
  app_number?: string | null
  applicant_a?: ApplicantSection | null
  applicant_b?: ApplicantSection | null
  other_income?: OtherIncomeSection | null
  reference?: ReferenceSection | null
  bank_reference?: BankReferenceSection | null
  authorization?: AuthorizationSection | null
  dealer_section?: DealerSection | null
  created_by?: string | null
  updated_by?: string | null
  submitted_by?: string | null
  submitted_at?: string | null
  created_at: string
  updated_at: string
  updated_by_name?: string | null
  submitted_by_name?: string | null
}

export interface CreditApplicationUpdatePayload {
  transaction_type?: TransactionType | null
  application_type?: ApplicationType | null
  app_number?: string | null
  applicant_a?: ApplicantSection | null
  applicant_b?: ApplicantSection | null
  other_income?: OtherIncomeSection | null
  reference?: ReferenceSection | null
  bank_reference?: BankReferenceSection | null
  authorization?: AuthorizationSection | null
  dealer_section?: DealerSection | null
  revert_to_draft?: boolean
}

export function emptyCreditApplication(): CreditApplicationUpdatePayload {
  return {
    transaction_type: null,
    application_type: "individual",
    app_number: null,
    applicant_a: {},
    applicant_b: {},
    other_income: {},
    reference: {},
    bank_reference: {},
    authorization: {},
    dealer_section: { vehicle: {}, trade_in: {} },
  }
}
